Kernel/Storage/SD/SDHostController.cpp

```diff
@@ -0,0 +1,667 @@
+            [&]() { return !(m_registers->present_state & command_inhibit); })) {
+        return EIO;
+    }
+
+    // 2. If the Host Driver issues an SD Command using DAT lines
+    //    including busy signal, go to step (3).
+    //    If without using DAT lines including busy signal, go to step (5).
+    // 3. If the Host Driver is issuing an abort command, go to step (5). In the
+    // case of non-abort
+    //    command, go to step (4).
+    if (cmd.requires_dat_line() && cmd.type != SD::CommandType::Abort) {
+
+        // 4. Check Command Inhibit (DAT) in the Present State register. Repeat
+        // this step until
+        //    Command Inhibit (DAT) is set to 0.
+        constexpr u32 data_inhibit = 1 << 2;
+        if (!retry_with_timeout(
+                [&]() { return !(m_registers->present_state & data_inhibit); })) {
+            return EIO;
+        }
+    }
+
+    // 5. Set registers as described in Table 1-2 except Command register.
+    m_registers->argument_1 = argument;
+
+    // 6. Set the Command register.
+    m_registers->transfer_mode_and_command = cmd.raw;
+
+    // 7. Perform Command Completion Sequence in accordance with 3.7.1.2.
+    // Done in wait_for_response()
+
+    return {};
+}
+
+ErrorOr<SDHostController::Response> SDHostController::wait_for_response()
+{
+    // SDHC 3.7.1.2 The Sequence to Finalize a Command
+
+    // 1. Wait for the Command Complete Interrupt. If the Command Complete
+    // Interrupt has occurred,
+    //    go to step (2).
+    if (!retry_with_timeout(
+            [&]() { return m_registers->interrupt_status & command_complete; })) {
+        return EIO;
+    }
+
+    // 2. Write 1 to Command Complete in the Normal Interrupt Status register to
+    // clear this bit
+    m_registers->interrupt_status = command_complete;
+
+    // 3. Read the Response register(s) to get the response.
+    Response r = {};
+    auto cmd = last_sent_command();
+    switch (cmd.response_type) {
+    case SD::ResponseType::NoResponse:
+        break;
+    case SD::ResponseType::ResponseOf136Bits:
+        r.response[0] = m_registers->response_0;
+        r.response[1] = m_registers->response_1;
+        r.response[2] = m_registers->response_2;
+        r.response[3] = m_registers->response_3;
+        break;
+    case SD::ResponseType::ResponseOf48Bits:
+        r.response[0] = m_registers->response_0;
+        break;
+    case SD::ResponseType::ResponseOf48BitsWithBusy:
+        // FIXME: Idk what to do here
+        break;
+    }
+
+    // 4. Judge whether the command uses the Transfer Complete Interrupt or not.
+    // If it uses Transfer
+    //    Complete, go to step (5). If not, go to step (7).
+    if (last_sent_command().uses_transfer_complete_interrupt()) {
+        // 5. Wait for the Transfer Complete Interrupt. If the Transfer Complete
+        // Interrupt has occurred, go to step (6).
+
+        while ((m_registers->interrupt_status & transfer_complete) == 0)
+            ;
```

Comment:
please call either one of `Processor::pause` or `Processor::wait_check` or `Processor::wait_for_interrupt`, unless you want the pi to catch fire

Reply:
I decided to remove this IF-branch and instead slap a TODO on it.

It's never reached and I'm not sure this function is fit to handle that type of commands anyway, I'd rather have someone fall into a TODO and find the issue quickly